README.md
sncr_mapr Cookbook
==========================
This cookbook automates the installation of a MapR cluster. Currently, this cookbook supports has only been tested on CentOS, but should work fine on Redhat as well

This cookbook will automate the steps outlined in the 'Preparing your nodes' section of the MapR documention, located at http://doc.mapr.com/display/MapR/Preparing+Each+Node, with the exception of verifying OS, mememory, etc.  This cookbook assumes that you have verified that the HW meets the requirements outline in the aforementioned link.

Note that this cookbook disables both selinux and iptables.  A list of MapR ports is available at http://doc.mapr.com/display/MapR/Ports+Used+by+MapR .

This cookbook will also automate the installation of a MapR Yarn cluster.


--------- install_mapr_chef.sh usage -----------------

NOTE:  The attached shell script automates the following steps:
  1.  Creates a log dir, ~/mapr_install_logs (if not already done)
  2.  Moves any previous run logs to ~/mapr_install_logs/bak
  3.  runs ssh <node_name> chef-client >><logdir_n_file> (and waits for complettion)
  4.  Prompts users to apply license, which can be done via the GUI (instructions will be provided during walk through), and waits for confirmation.
  5.  Shuts down mapr-warden services and reboots all nodes, to allow automount to be enabled on the cluster.
  6.  Restarts all mapr-wardens once more to ensure clean service state for all mapr services.

The only manual step for this script is updating the 'nodes' value at the top of the script to match the FQDN of the nodes to be included in the MapR cluster.
---------end  install_mapr_chef.sh usage -------------

Requirements
------------
This cookbook requires the ntp cookbook located at https://supermarket.chef.io/cookbooks/ntp.

This cookbook also comes with a wrapper script, called 'install_mapr_chef.sh'.
This cookbook will install all SW dependencies for MapR, so nothing to do there.

Attributes
----------
There are a number of configurable attributes, all located in attributes/default.rb.  The most significant ones (as well as the ones users MUST configure) as listed here.

The below is a list of all of the nodes that will be part of the cluster.  The below is an example of nodes located in EC2.  NOTE:  You MUST use FQDN for this list and all other values in attributes.

```ruby
default[:mapr][:cluster_nodes] =
  [ "ip-172-16-5-225.ec2.internal",
     "ip-172-16-5-16.ec2.internal",
	 "ip-172-16-5-176.ec2.internal",
	 "ip-172-16-5-108.ec2.internal",
	 "ip-172-16-5-37.ec2.internal",
	 "ip-172-16-5-79.ec2.internal"
  ]
```

****NOTE:  There is a similar list located in the attached shell script, 'mapr_install_chef.sh', which will need to be configured as well.


Total node count to be installed:
```ruby
default[:mapr][:node_count] = "6"

default[:mapr][:cldb] = ["ip-172-16-5-16.ec2.internal","ip-172-16-5-176.ec2.internal"]
default[:mapr][:zk] = ["ip-172-16-5-108.ec2.internal","ip-172-16-5-37.ec2.internal","ip-172-16-5-79.ec2.internal"]
default[:mapr][:rm] = ["ip-172-16-5-225.ec2.internal","ip-172-16-5-16.ec2.internal"]
default[:mapr][:hs] = "ip-172-16-5-225.ec2.internal"
default[:mapr][:ws] = ["ip-172-16-5-225.ec2.internal","ip-172-16-5-16.ec2.internal"]
```

The name you would like for your cluster and the version to install. 4.0.2 is the current GA level, so shouldn't need to be changed.
default[:mapr][:clustername] = "chef_test_cluster"
default[:mapr][:version] = "4.0.2"

The disks MapR should use for installation.  These MUST be raw, unformatted drives.  A list can be confirmed w/ lsblk.
```ruby
default[:mapr][:node][:disks] = "/dev/xvdf,/dev/xvdg"
```

Java version to be installed as well as home directory for setting environment and mapr-specific environment variables.
```ruby
default[:java][:version] = "java-1.7.0-openjdk-devel"
default[:java][:home] = "/usr/lib/jvm/jre-1.7.0-openjdk.x86_64"
```



Usage
-----
#### sncr_mapr::default

Assuming the above is configure, just include `sncr_mapr` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[sncr_mapr]"
  ]
}
```

You will typically have to override the various node 'function' assignments and instance lists in the environment file.

In addition, you will need a databag for the mapr user.  The default databag is 'users', but there is an attribute to allow setting a users data bag on a per environment basis.  Here is a typical entry in a databag for a mapr user.

```json
{
  "id": "mapr",
  "password": "$1$x4GGdWjX$5IZsb6wdcgGjzdSKK6yd6/",
  "ssh_keys": [ "ssh-rsa ... == root@node1"   ],
  "groups": [ "mapr" ],
  "shell": "/bin/bash",
  "comment": "MapR cluster user"
}
```

## Ecosystem Components

This cookbook includes recipes to install some of the hadoop ecosystem components.  These should be used with the chef 'role' methodology. I.e., creating chef roles, and assigning those roles to the various instances after the core cluster is running.

The following recipes can be assigned to roles to add these ecosystem components.

## Spark

* ::spark_master
* ::spark_worker

## Drill

* ::drill_worker

## Oozie

* ::oozie_server
* ::oozie_client

Contributing
------------

License and Authors
-------------------


CHANGELOG.md
sncr_mapr CHANGELOG
===================

This file is used to list changes made in each version of the sncr_mapr cookbook.

## v0.13.0

* drill configs

## v0.12.0

* spark configs
  * make spark-env a tempate
  * set memory params from attrs
  * do not make master a worker by default

## v0.11.0, v0.11.1

* improve config change detection
* make ecosystem yum repo major version specific

## v0.10.0, v0.10.1, v0.10.2

* update the spark-env.sh to include the public_dna
* try to stop/start slaves when this happens
* add a not_if in case the spark_master has not been created yet
* fix ruby error (missing {} )

## v0.9.1

* in spark_master, protect the reconfigure so it doesn't happen all the time


## v0.9.0

* customize ~mapr/.bashrc

## v0.8.3

* fix missing _ in node index

## v0.8.2

* fix attribute error, missing w in %w

## v0.8.1

* add -p to the -mkdir for /apps/spark

## v0.8.0

* rename attributes
* simplify prereq package loading


## v0.7.0

* rename to sncr_mapr
* remove iptables and selinux recipes
* move their disables to install_pre...


## v0.6.2

* set the core-site.xml action and notifies

## v0.5.0

* fix bug in ::spark_master
## v0.5.0

* install drill

## v0.4.0

* install Apache Spark

## v0.3.2

* add the rpm gpgkey
* turn more ruby blocks into bash blocks

## v0.3.1

* make the /etc/clustershell directory
* this version successfully launched a single node cluster

## v0.3.0

* switch to java cookbook
* some fixes

## v0.2.1

* don't blast root's ssh keys and config


## v0.2.0

0.1.0
-----
- [your_name] - Initial release of sncr_mapr

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


install_mapr_chef.sh
#!/bin/bash

#Node list:
nodes="ip-172-16-2-225.ec2.internal ip-172-16-2-16.ec2.internal ip-172-16-2-176.ec2.internal ip-172-16-2-108.ec2.internal ip-172-16-2-37.ec2.internal ip-172-16-2-79.ec2.internal"
log_date=`date +%F_%H-%M`

if [ ! -d ~/mapr_install_logs ]; then
	echo "Making mapr_install directory"
	mkdir -p ~/mapr_install_logs/bak
else
  if [[ "`ls ~/mapr_install_logs/*.install.log 2>/dev/null`" != "" ]]; then
	echo "Moving old installation logs"
	mv ~/mapr_install_logs/*.install.log ~/mapr_install_logs/bak
  fi
fi

for i in $nodes; do
	echo "Starting chef-client run for node $i"
	nohup ssh $i chef-client >>~/mapr_install_logs/$i.$log_date.install.log 2> ~/mapr_install_logs/error.$i.$log_date.install.log < /dev/null &
done

sleep 10 

for i in $nodes; do
	while [[ "`ssh $i ps -ef|grep chef-client|grep -v grep|grep -v ssh`" != "" ]]; do
	  sleep 20;
          echo "Waiting for $i to finish chef-client"
	done
done

applied="no"
while [[ "$applied" != "y" ]]; do 
  echo -n "Have you applied a MapR license ('y' if so, 'q' to quit script):  "
  read applied
  case $applied in
    q*)
      echo -e "\n\nQuitting script...\n\n"
      exit
  esac
done

echo -e "\n\n###IMPORTANT!!  THIS NEXT STEP WILL REBOOT THE SERVERS###"
echo -e "###IMPORTANT!!  THIS NEXT STEP WILL REBOOT THE SERVERS###\n\n"

rb="no"
while [[ "$rb" != "y" ]]; do 
  echo -n "Reboot all servers?('y' if so, 'q' to quit script):  "
  read rb
  case $rb in 
    q*)
      echo "Quitting script..."\n\n
      exit
  esac
done 

echo -e "\n\nRebooting all servers!!!\n\n"
echo -e "Waiting for all servers to come back\n\n"

for i in $nodes; do
  if [[ "`ssh $i ls /opt/mapr/roles|grep cldb`" != "cldb" ]]; then 
    ssh $i service mapr-warden stop
  fi
done

for i in $nodes; do 
  if [[ "`ssh $i ls /opt/mapr/roles|grep cldb`" == "cldb" ]]; then 
    ssh $i service mapr-warden stop
  fi
done

for i in $nodes; do
  if [[ "`ssh $i ls /opt/mapr/roles|grep zookeeper`" == "zookeeper" ]]; then
    ssh $i service mapr-zookeeper stop
  fi
done

for i in $nodes; do
  ssh $i reboot
done

for i in $nodes; do
  while [[ "`ssh $i uname -m 2>/dev/null`" != "x86_64" ]]; do
    sleep 20
    echo "Sleeping 20 seconds for host $i"  
  done 
done


node_count=`echo $nodes|wc -w`
echo "node_count == $node_count"
warden_up="0"
echo "Nodes = $nodes"
for i in $nodes; do 
  while [[ "`ssh $i service mapr-warden status`" != "WARDEN running as process "[0-9]*\. ]]; do 
    sleep 5
    echo "sleeping for $i warden"
 done
    echo "Warden on $i is  up"
    warden_up+=1
done


test=`echo $nodes|awk '{print $1}'`
#echo -e "\n\ntest == $test\n\n"

echo -e "\n\nLooking for active CLDB\n\n"
while [[ "`ssh $test maprcli node cldbmaster|awk '{print $1}'`" != "cldbmaster
ServerID:" ]]; do
  sleep 5
  echo -e "Waiting for active CLDB..."
done

echo -e "\n\nFound active CLDB!\n\n"

echo -e "Restarting all Wardens\n\n"
for i in $nodes; do
  ssh $i service mapr-warden restart
done

test=`echo $nodes|awk '{print $1}'`
#echo -e "\n\ntest == $test\n\n"

echo -e "\n\nLooking for active CLDB\n\n"
while [[ "`ssh $test maprcli node cldbmaster|awk '{print $1}'`" != "cldbmaster
ServerID:" ]]; do
  sleep 5
  echo -e "Waiting for active CLDB..."
done

echo -e "\n\nFound active CLDB!\n\n"

echo -e "\n\n\nAll Done Here!!!!!\n\n\n"
